keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,297 @@
+  section][] for the schema) if it doesn't exist already.
+* for each newly created `migration` instance, updates the
+  `migration.spec.lastUpdateTimestamp`, and logs the reason.
+* the `ownerReferences` of the `migration` objects are set to the **kube-migrator
+  controller** `deployment`. Thus, when user invokes the **kube-migrator
+  install** command, the old `migration`s are deleted with the old `deployment`.
+
+The control loop of **kube-migrator controller** does the following:
+* watches the instances of the `migration` CR.
+* if `migration.metadata.generation != migration.status.observedGeneration`
+    * deletes existing **migration worker** `job` for the resource type.
+    * launches new **migration worker** `job` to migrate the resource type. The
+      `job` is annotated with the generation.
+    * makes sure there is only one **migration worker** `job` running at a time,
+      to avoid overloading the apiserver.
+    * updates `migration.status.observedGeneration`.
+* watches the **migration worker** jobs. Updates
+  `migration.status.completedGeneration` when a job completes.
+* short-circuit: if `migration.spec.groupVersions` shows that the resource has
+  only one version, then its `.status.observedGeneration` and
+  `.status.completedGeneration` are updated directly without launching the
+  worker `job`.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in a `configmap`
+to preserve progress. With the inconsistent continue token introduced in
+[#67284][], the **migrator worker** does not need to worry about expired continue
+token.
+
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#api
+
+The cluster admin can run the `kube-migrator progress` command to check if the
+migration has completed. The command lists all `migrations` instances and checks
+if all of them have `.spec.generation == .status.observedGeneration`.
+
+### API
+
+We introduce the `migration` API to record the intention and progress of a
+migration. The API will be a CRD defined in the `migration.k8s.io` group.
+
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  Spec StorageVersionMigrationSpec
+  Status StorageVersionMigrationStatus
+}
+
+// Note that the apiserver will increment `metadata.generation` when this spec
+// is updated.
+type StorageVersionMigrationSpec {
+  // Resource is the identifier of the resource that's being migrated.
+  // TBD: maybe resource=hash(etcd instance, etcd path prefix).
+  // Immuntable.
+  Resource string 
+
+  // GroupVersions are all the groups and versions the apiserver serves this
```

Comment:
Why is it necessary to pass all? Shouldn't a single one be sufficient? And why do you need to care whether it is preferred or not?

Reply:
> Why is it necessary to pass all?

For optimization. The migrator can skip the resource if this resource only has one version.

In the iteration addressing your comments, I updated the KEP to say that at the alpha stage, the tool migrates a hard-coded list of resources, this field became unnecessary.

> And why do you need to care whether it is preferred or not?

I thought if the migrator GET&PUT at the preferred version could save the apiserver a few conversions, but my calculation was wrong.